### ISSUE TITLE
Decimal or fraction support

### ISSUE THREAD

**Issue by meisyal** [nlp-compromise/nlp_compromise#126]
First of all, I would like to say congratulation for making such amazing JavaScript NLP library.

This library is great because it's my first time seeing NLP library that can detect a number or even a unit from a term. For instance, I have a term `five kilometers`. This library can detect a number (`five` to be `5`) and a unit (`kilometers` to be `kilometer` or `km`). But, this library doesn't support decimal or fraction in a term. That would be great if this library supports it. For example, I have a term `one-half teaspoon of sugar`. It can be detected by this library, `one-half` to be `1.5`.

**Comment by jashsayani** [comment]
This is interesting, as decimal detection is locale sensitive. So in `en-us`, we have 1.45 whereas in `fr-fr` we have 1,45 (use of `,` over `.` as a decimal character).

**Comment by spencermountain** [comment]
thanks Andrias, I was surprised about this too.
great idea. Potentially it could handle `one and two-fifths teaspoons` as well.
I once tried handling "five and a half thousand", etc but it got a little sloppy. Not impossible.

@jashsayani ooof! this is true isn't it. shiiiiiit. Is there a way to mitigate it? I suppose it's less-ambiguous when there's less than three digits after the comma? ... i'm not sure.... wow! ;).

**Comment by jashsayani** [comment]
>  Is there a way to mitigate it?

Yes, I would suggest having a normalized value that is used for processing. While the display value would have `,` or `.` based on locale, this can be normalized to always be `.` in memory where it is processed, and the displayed value can be passed through internationalization to display the appropriate separators.

**Comment by scagood** [comment]
Here are some possible thoughts.

Example table with decimals/separators.
```
The simple ones:
100,000,231.237 -> Separator: ',', Decimal: '.'
1,373.432       -> Separator: ',', Decimal: '.'
100.000.231,237 -> Separator: '.', Decimal: ','
100 000 231,237 -> Separator: ' ', Decimal: ','
1.373,432       -> Separator: '.', Decimal: ','
100000.2353     -> Separator: '',  Decimal: '.'
100000,2353     -> Separator: '',  Decimal: '.'
1.234.234       -> Separator: '.', Decimal: ''
1,234,234       -> Separator: ',', Decimal: ''
1 234 234       -> Separator: ' ', Decimal: ''
.23             -> Separator: '',  Decimal: '.'
1,45            -> Separator: '',  Decimal: ','

The 'hard' ones:
These could be as follows:
123.321         -> Separator: '.', Decimal: ''
123,321         -> Separator: ',', Decimal: ''

Or like this:
123.321         -> Separator: '',  Decimal: '.'
123,321         -> Separator: '',  Decimal: ','
```

So, as a table doesn't help much in the scale of things, here is my 'crack' at thinking this through:

---

First, detecting the format form the numbers with both separators present could be done using a 'simple' regex expression like so (assuming it's already been split as a term):
```js
const bothSepDec = /^\d+([ ,.])(?:\d+\1?)+(?!\d)((?!\1)[,.])\d+$/;
```

This will detect the following:
 * 100,000,231.237
 * 1,373.432
 * 100 000 231,237
 * 1.373,432

In addition, the matched groups will be the separators, e.g:
```js
output = '100,000,231.237'.match(/^\d+([ ,.])(?:\d+\1?)+(?!\d)((?!\1)[,.])\d+$/);

// Output ["100,000,231.237", ",", "."]
```

---

Next we can find the numbers with multiple separators of the same type:
```js
const multiSep = /^\d+([ ,.])\d+(?:\1\d+)+$/;
```

This will detect the following:
 * 100,000,231,237
 * 1,373,432
 * 1 234 234
 * 1,234,234

Again the matched group would be the separator:
```js
output = '100,000,231'.match(/^\d+([ ,.])\d+(?:\1\d+)+$/);

// Output ["100,000,231", ","]
```

---

Next, we can simply detect the smaller decimals, as it's unlikely to be a separator: 
```js
const smallDec = /^\d*([,.])\d{0,2}$/;
// This could be `/^\d*([,.])(?:\d{0,2}|\d{4,})$/` if we're just avoiding multiples of 3?
```

This should quickly clear up the more common `1,45€` example.

---

I'm completely unsure how to deal with triple decimal places like: `1.237` or `23,856`?
Possibly the only way is to know the locality before parsing? 

This is the site I used as a reference [syllabus.ca/.../writing-numbers-in-french-and-english/](http://www.syllabus.ca/en/didyouknow-en/writing-numbers-in-french-and-english/) and it seems to imply a space could also be a numerical separator? (I initially thought this `123.412.234,123` was a plausible format, however, the page made me amend my thinking to; `123 412 234,123` and `123412234,123`, however, I included all three in my musings :) ).

This unfortunately leads to another issue, if we try to parse a `fr-fr` number in the thousands, lets say; `12 234,12` it will see two separate values.
```js
nlp('12 234,12')
    .values()
    .debug();

// Outputs:
// ====
//    --
//    '12'                       - NumericValue, Value, Cardinal
//    --
//    '234,12'                   - NumericValue, Value, Cardinal
```

Leading to the question, should this be compromise's job or should this be for [the french version of compromise](https://github.com/nlp-compromise/fr-compromise)?

**Comment by spencermountain** [comment]
@scagood you are my new favourite person.
I support this completely.

**Comment by spencermountain** [comment]
yeah, lets support the all the unambiguous formats we can, and default to an english-understanding for the ambiguous forms. fr-compromise can do whatever it wants to do. that stuff is well-separated.
